external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MsSqlServerIntegrationSuite.scala

```diff
@@ -59,7 +59,7 @@ class MsSqlServerIntegrationSuite extends DockerJDBCIntegrationSuite {
       """
         |INSERT INTO numbers VALUES (
         |0,
-        |255, 32767, 2147483647, 9223372036854775807,
+        |127, 32767, 2147483647, 9223372036854775807,
```

Comment:
Ditto - https://github.com/apache/spark/pull/26549#discussion_r347202275. Can anyone explain why it was possible, and how do we handle unsigned cases?

Reply:
Continuing from https://github.com/apache/spark/pull/26549#discussion_r347213679 : TINYINT looks like it's a single byte alright, so using ByteType is reasonable. However it looks like it's treated as signed by some but not all DBMSes. Is it unsigned in SQL Server?

Just checking: these types like TINYINT and SMALLINT are not standard types, although widely supported, right? should these types be used by default for all JDBC sources?

Yeah I have some more doubts now that the TINYINT issue was pointed out. @shivsood

Reply:
Got it. @gatorsmile , @HyukjinKwon , @srowen . I overlooked that mismatch between TINYINT vs Byte in this PR.

Reply:
`SMALLINT` is widely supported because it's SQL-92. For TINYINT, I agree that we need to revert partially for that type.

Reply:
Thanks all for pointing out these issues. I had overlooked handling of unsigned cases and the fact that each database may define on its own. I think the problem exists for both SMALLINT and TINYINT. 
- TINYINT is very clear that it can be range any where from -127 to +127 and unsigned as 0 to 255. Both [SQLServer](https://docs.microsoft.com/en-us/sql/t-sql/data-types/int-bigint-smallint-and-tinyint-transact-sql?view=sql-server-ver15) and [MySQL] (https://dev.mysql.com/doc/refman/8.0/en/integer-types.html)
- SMALLINT can take an unsigned value 0 to 65535 per [MySQL] (https://dev.mysql.com/doc/refman/8.0/en/integer-types.html) 

My understanding is that there are no unsigned type in Spark. c.f. https://spark.apache.org/docs/latest/sql-reference.html. Is that assertion right?

Do we have a test for an integer where integer value is 4294967295. Per [MySQL] (https://dev.mysql.com/doc/refman/8.0/en/integer-types.html)  documentation that's possible that an unsigned integer will have that value.

Reply:
Thanks guys for addressing my comment thoroughly and taking care of this afterwards.

Reply:
So at this stage the only solution i have for this problem is 
1. Map SMALLINT as IntegerType
2. Map TINYINT as ShortType

and ofcourse add test cases to cover all signed and unsigned ranges. I can make these changes is very one is ok with this plan?

@HyukjinKwon @gatorsmile @dongjoon-hyun @srowen @maropu Thanks for your comments. Please let me know if you are ok with the plan above.

Reply:
looks fine to me.

Reply:
Since some dbmss don't support unsigned cases, I just want to know the exact behaviour first that we should support from Spark to the all the existing dbmss and vice versa.

Reply:
The behavior would be as follows.
### Overwrite scenario
1. (ShortType -> Int) If a spark df has a ShortType, on overwrite a DBMSS table with type Int will get created. Because Spark ShortyType is -32768 to +32767 only these values can be written. 
2. (ByteType -> SmallInt) In a spark df has a ByteType, on overwrite  DBMSS table with type Short will get created. Because Spark ShortyType is -128 to +127 only these values can be written. 

### Read scenario
1. If an existing table in DBMSS has type TinyInt, a read in Spark would results in a ShortType. Because ShortType range in Spark in -32786 to +32768, DBMSS signed value -127 to +127 and unsigned range of 0 to 255 will be handled.

2. If an existing table in DBMSS has type SmallInt, a read would result in Spark dataframe having a column type as Integer. Because Integer range in Spark in -2147483648 to +2147483648, DBMSS signed value --32768 to +32768 as well as unsigned range of 0 to 65535 will be handled.

Reply:
So we have to use a widening conversion both ways. That's the safest thing to do, I guess, and less of a change from the current behavior, where bytes go all the way to ints.

Reply:
Ur, I see. right.

Reply:
Anyway, thanks for the explanation, @shivsood

Reply:
Raised PR https://github.com/apache/spark/pull/27172 with the proposed fix. ShortType is unchanged, only ByteType is fix is modified to map to ShortType on the read path so enable support for  0 to 255 range. 

@maropu @srowen @HyukjinKwon  as FYI. Thanks